chore: add Braze migration tutorial (#1044)
knocklabs/docs: content/tutorials/migrate-from-braze.mdx

---
title: Migrate from Braze to Knock
description: Learn how to migrate your notifications from Braze to Knock.
tags: ["migrate", "braze", "migration"]
section: Tutorials
---

Knock's APIs and developer tools make it easy to migrate your notification templates and user data from other notifications platforms into Knock. In this tutorial, we will walk you through planning and executing a migration from Braze into Knock, focusing on transactional messaging workflows.

## Mapping Braze concepts to Knock concepts

Before migrating any data into Knock, it's helpful to understand how the resources in your Braze account map to concepts and resources in Knock.

### Integrations

In Braze, you configured <a href="https://www.braze.com/docs/user_guide/message_building_by_channel" target="_blank">messaging channels</a> (email, SMS, push, etc.) to deliver notifications through various providers. In Knock, we refer to these delivery platforms as [channels](/concepts/channels).

Channels are configured under the **Integrations** tab in your Knock dashboard. You can see a full list of supported channel types and providers [here](/integrations/overview).

> In addition to first-party integrations with message delivery platforms, Knock also offers convenient connections to customer data platforms (CDPs) and reverse ETL providers to bring your data into Knock ([sources](/integrations/sources/overview)), as well as to popular analytics and data warehousing tools to enable you to export important data out of Knock ([extensions](/integrations/extensions/overview)).

Author: Won't action ([see comment](https://github.com/knocklabs/docs/pull/1044#issuecomment-3108790254))

### Users

Braze uses <a href="https://www.braze.com/docs/user_guide/getting_started/users_segments/#users" target="_blank">users</a> to represent the recipients of notifications, with custom attributes and profile data associated with each user.

> Knock uses a similar concept with [users](/concepts/users) objects on which you can store any number of custom properties related to your notifications' recipients.

Author: Won't action ([see comment](https://github.com/knocklabs/docs/pull/1044#issuecomment-3108790254))

### Campaigns and workflows

> In Braze, your transactional messaging is handled through <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/email/transactional_message_api_campaign" target="_blank">transactional email campaigns</a>, <a href="https://www.braze.com/docs/user_guide/engagement_tools/campaigns/building_campaigns/delivery_types/api_triggered_delivery/" target="_blank">API-triggered campaigns</a>, and <a href="https://www.braze.com/docs/user_guide/engagement_tools/canvas/" target="_blank">canvases</a>. These campaigns contain message templates and are triggered via API calls using a `campaign_id`.

Author: Thanks! Will reflect that in our docs for consistency.

Knock combines these into a single resource called a [workflow](/concepts/workflows), which serves as a container for all of the logic and message templates associated with a given notification in your system.

When you’re ready to start sending notifications, you’ll do so by [triggering](/send-notifications/triggering-workflows) your workflows, similar to how you trigger Braze campaigns.

### Template management

> Braze uses <a href="https://www.braze.com/docs/user_guide/engagement_tools/templates_and_media/content_blocks" target="_blank">content blocks</a> for reusable content and Liquid templating for personalization within campaigns.

Author: Similar to Canvases, since Braze always references "Content Blocks" as a proper noun, I will reflect that in our docs for consistency.

Knock provides similar functionality through [partials](/concepts/partials) for reusable content blocks and also supports Liquid templating for personalization within workflow templates.

### Multi-language support and translations

> Braze provides <a href="https://www.braze.com/docs/user_guide/administrative/app_settings/multi_language_settings/" target="_blank">multi-language settings</a> that enable you to target users in different languages and locations with different messages within a single email campaign. Braze keeps every translation file inside the campaign (or canvas/email template) that uses it. You wrap the text in `{%translation%}` tags, export a locale CSV, fill it in, and re‑upload each time that specific campaign's copy changes.

Author: This is where I found the CSV steps: https://www.braze.com/docs/user_guide/message_building_by_channel/email/using_locales

I'll update to also mention the API 👍

Knock helps you to power notifications in multiple locales and languages using [translations](/concepts/translations). You can work with translations directly in your dashboard or programmatically via API, and Knock supports both `json` and `.po` file formats.

> When using the `t` tag [method](/concepts/translations#translation-methods-filter-vs-tag) of referencing translations in your message templates, Knock will automatically generate the associated translation files for each of your registered locales behind the scenes.

Author: Love this

### Multi-tenancy

In Braze, there isn't a direct equivalent to Knock's tenant functionality. Most Braze customers handle customer/organization segmentation through custom attributes (like `organization_id`, `account_id`, `company_name`) combined with segments built on those attributes. This approach requires manual campaign targeting, complex segmentation logic, and offers no native support for per-organization branding or scoped in-app feeds.

> In Knock, [tenants](/concepts/tenants) provide native multi-tenancy support that eliminates these workarounds. Tenants represent segments your users belong to—what you might call "accounts," "organizations," "workspaces," or similar. Per-tenant branding attributes are stored directly on a tenant in Knock rather than requiring separate campaigns or complex attribute management. Knock tenants are applied as context to workflow triggers to automatically [apply per-tenant branding](/concepts/tenants#custom-branding), [manage per-tenant preferences](/preferences/tenant-preferences), and [scope in-app feed messages to particular tenants](/concepts/tenants#scoping-in-app-feeds).

Author: For reference, Matt and I chatted offline about this. I pulled this in from https://docs.knock.app/concepts/tenants, but I'm going to use Matt's suggestion, as I think that provides better clarity. We will add a project to review the tenants page in more detail to identify opportunities for providing better clarity for users new to Knock.

          </a>
      </Accordion>
    </AccordionGroup>

  </Step>
  <Step title="Migrate translations">
    If you're using multi-language settings in Braze, you can extract your translation content and import it into Knock's translation system.

    Key considerations to remember:
      - In Braze, translation files are stored at a campaign level.
      - In Knock, translations are managed globally at the environment level, and are reusable across workflows.

    Suggested approach:
      1. Export a list of all Braze campaigns, canvases and email templates that include translation tags and note which locales they cover.
      2. Use Braze’s bulk‑translation APIs to pull down the raw locale data for those assets.
      3. Each Braze export comes back grouped by message variant. You will want to decide on a universal key pattern for Knock and convert the Braze JSON/CSV into locale‑key‑value maps that match this pattern. Be sure to check pluralization, as Braze sometimes stores counts as separate rows, but Knock supports [zero/one/other rules](/concepts/translations#pluralization) out‑of‑the‑box.

>       4. For each locale, call Knock’s Management API to upsert the file into your Knock environment. Once the locale is in place, templates can reference the strings with the `t` tag.

Author: Ope yup - thanks for catching that! 😅

      <Accordion title="Relevant endpoints">
          <a href="https://www.braze.com/docs/api/endpoints/translations/campaigns/get_bulk_translations_campaigns" target="_blank" style={{textDecoration: 'none'}}>
            <Endpoint method="GET" path="https://rest.{instance}.braze.com/campaigns/translations" />
          </a>
           <a href="https://www.braze.com/docs/api/endpoints/translations/canvas/get_bulk_translations_canvases" target="_blank" style={{textDecoration: 'none'}}>
            <Endpoint method="GET" path="https://rest.{instance}.braze.com/canvas/translations" />
          </a>
           <a href="https://www.braze.com/docs/api/endpoints/translations/email_templates/get_view_translation_template" target="_blank" style={{textDecoration: 'none'}}>
            <Endpoint method="GET" path="https://rest.{instance}.braze.com/templates/email/translations" />
          </a>
          <a href="https://docs.knock.app/mapi-reference/translations/upsert" target="_blank" style={{textDecoration: 'none'}}>
            <Endpoint method="PUT" path="https://control.knock.app/v1/translations/{locale_code}" />
          </a>
      </Accordion>
    </AccordionGroup>

  </Step>

>   <Step title="Add tenants (if applicable)">

Author: Agreed - was on the fence with that as well 😅

>     If your Braze implementation uses custom attributes to segment users by organization, account, or workspace, you can migrate this data to create proper tenant structures in Knock. This migration will replace complex attribute-based segmentation with Knock's native tenant functionality.

Author: Yeah I played around with this, and I think "associated users with" provides better clarity. 👍

---
title: Migrate from Braze to Knock
description: Learn how to migrate your notifications from Braze to Knock.
tags: ["migrate", "braze", "migration"]
section: Tutorials
---

Knock's APIs and developer tools make it easy to migrate your notification templates and user data from other notifications platforms into Knock. In this tutorial, we will walk you through planning and executing a migration from Braze into Knock, focusing on transactional messaging workflows.

> ## Mapping Braze concepts to Knock concepts

Reviewer: Table could be nice. Maybe just separate from this content as a quick reference or something. Hard part there will be keeping it readable in two column format within our docs page where reading window is already fairly narrow. Also need to consider mobile. 

Let's ship this as-is and can consider a table component further down the line 👍

Channels are configured under the **Integrations** tab in your Knock dashboard. You can see a full list of supported channel types and providers [here](/integrations/overview).

In addition to first-party integrations with message delivery platforms, Knock also offers convenient connections to customer data platforms (CDPs) and reverse ETL providers to bring your data into Knock ([sources](/integrations/sources/overview)), as well as to popular analytics and data warehousing tools to enable you to export important data out of Knock ([extensions](/integrations/extensions/overview)).

### Users

Braze uses the concept of <a href="https://www.braze.com/docs/user_guide/getting_started/users_segments/#users" target="_blank">users</a> to represent the recipients of notifications, with custom attributes and profile data associated with each user.

Knock uses a similar concept of [users](/concepts/users) objects on which you can store any number of custom properties related to your notifications' recipients.

### Campaigns and workflows

> In Braze, your transactional messaging is handled through <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/email/transactional_message_api_campaign" target="_blank">transactional email campaigns</a>, <a href="https://www.braze.com/docs/user_guide/engagement_tools/campaigns/building_campaigns/delivery_types/api_triggered_delivery/" target="_blank">API-triggered campaigns</a>, and <a href="https://www.braze.com/docs/user_guide/engagement_tools/canvas/" target="_blank">Canvases</a>. These campaigns contain message templates and are triggered via API calls using a `campaign_id`.

Reviewer 2: > are triggered via API calls using a `campaign_id`.

Braze has other trigger options also. Maybe best to just remove this sentence since the target audience should be generally familiar with what is possible in Braze.

Knock combines these into a single resource called a [workflow](/concepts/workflows), which serves as a container for all of the logic and message templates associated with a given notification in your system.

> When you’re ready to start sending notifications, you’ll do so by [triggering](/send-notifications/triggering-workflows) your workflows, similar to how you trigger Braze campaigns.

Reviewer 2: ```suggestion
When you’re ready to start sending notifications, you’ll do so by [triggering](/send-notifications/triggering-workflows) your workflows, similarly to how you trigger Braze campaigns.
```

Key advantages of Knock's tenant approach:

- **Single workflow, multiple tenants.** One workflow can serve all organizations with tenant-specific customizations.
- **Native branding support.** Per-tenant logos, colors, and styling without template duplication.
- **In-app feed scoping.** In-app notifications can be filtered by tenant context.
- **Simplified preference management.** Per-user, per-tenant preferences without complex attribute juggling.

<Callout
  emoji="✨"
  title="Note:"
  text={
    <>
      Per-tenant branding and per-tenant preferences are features of our{" "}
      <a href="https://knock.app/pricing" target="_blank">
        Enterprise plan
      </a>
      . If you’d like to find out more information about Enterprise plan features
      and pricing, please contact us at <a href="mailto:sales@knock.app">
        sales@knock.app
      </a>.
    </>
  }
/>

### Subscriptions and preferences

> Braze manages user subscription preferences through <a href="https://www.braze.com/docs/api/endpoints/subscription_groups" target="_blank">subscription groups</a> and global subscription states (`email_subscribe` and `push_subscribe`.) Users can be subscribed, unsubscribed, or opted-in to different messaging channels and specific subscription groups. See Braze's documentation on <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/email/managing_user_subscriptions" target="_blank">email subscription management</a> and <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/sms_mms_rcs/subscription_groups" target="_blank">SMS subscription groups</a> for more details.

Reviewer 2: ```suggestion
Braze manages user subscription preferences through <a href="https://www.braze.com/docs/api/endpoints/subscription_groups" target="_blank">subscription groups</a> and global subscription states (`email_subscribe` and `push_subscribe`). Users can be subscribed, unsubscribed, or opted-in to different messaging channels and specific subscription groups. See Braze's documentation on <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/email/managing_user_subscriptions" target="_blank">email subscription management</a> and <a href="https://www.braze.com/docs/user_guide/message_building_by_channel/sms_mms_rcs/subscription_groups" target="_blank">SMS subscription groups</a> for more details.
```

Knock provides two features that together replace Braze subscription groups:

> - [Preferences](/preferences/overview) handle communication opt-outs similar to Braze's marketing-focused subscription groups, with enhanced flexibility for channel-specific, category-based, or workflow-specific preferences that can be [tenant-specific](/preferences/tenant-preferences) or [conditional](/preferences/preference-conditions).

Reviewer 2: ```suggestion
- [Preferences](/preferences/overview) handle communication opt-outs similarly to Braze's marketing-focused subscription groups, but with enhanced flexibility for channel-specific, category-based, or workflow-specific preferences that can be [tenant-specific](/preferences/tenant-preferences) or [conditional](/preferences/preference-conditions).
```

- Braze subscription groups for specific entities (project alerts, account updates) → Knock object subscriptions

## Migrating your data into Knock

Now that you have a good understanding of how the resources in your Braze account map to concepts and resources in Knock, you can start planning your migration.

Knock offers APIs and developer tools that make a migration smooth and efficient:

- A [Management API](/mapi) that enables you to work programmatically with the resources that you can also create directly in your Knock dashboard (like workflows and their associated message templates, email [layouts](/integrations/email/layouts), and translations.)
- A command line interface ([Knock CLI](/developer-tools/knock-cli)) that wraps the Management API, enabling you to work with your dashboard resources from the command line.
- [Bulk endpoints](/api-reference/overview/bulk-endpoints) that enable you to upsert large amounts of data in a single API request (more on specific endpoints below.)
- [Knock MCP server](/developer-tools/mcp-server) that enables AI assistants to help migrate workflows and templates from other platforms.

> <Callout

Reviewer 2: IMO, this callout feels out of place. Could we work this into the Migrate users step?

Author: @scoti-knock I moved this callout to the migrate users step (and then removed the third bullet point that essentially said the same thing). Does this feel better to you? https://docs-git-rt-migrate-from-braze-knocklabs.vercel.app/tutorials/migrate-from-braze#migrate-users

Reviewer 2: @rachael-t yes, I think that looks great!  🙌

  emoji="🚸"
  title="Inline recipient identification."
  text={
    <>
      While the following steps outline a suggested order for migrating
      individual resource types into Knock based on your existing Braze
      integration, it’s helpful to note that Knock also supports{" "}
      <a href="/managing-recipients/identifying-recipients#inline-identifying-recipients">
        inline identification
      </a>{" "}
      of recipients in order to enable you to upsert recipients as you are
      performing other actions like triggering a workflow or creating
      subscriptions. Your approach may vary depending on your specific
      requirements.
    </>
  }
/>

We recommend migrating data into Knock in the following order to ensure that certain resources which are dependencies of other resources are migrated first:

<Steps titleSize="h3">
  <Step title="Configure your integrations">

>     You'll want to configure your channels prior to migrating any workflows so that you can set the correct delivery methods for each of your notifications.

Reviewer 2: ```suggestion
You'll want to configure your [channels](/concepts/channels) prior to migrating any workflows so that you can set the correct delivery methods for each of your notifications.
```
